### PR TITLE
feat(vm): implement missing BigInt functions

### PR DESCRIPTION
**Pull request by load1n9**

### CODE REVIEW

File: nova_vm/src/ecmascript/types/language/bigint.rs [hunk 1]
```diff
@@ -232,35 +232,144 @@ impl BigInt {
     ///
     /// The abstract operation BigInt::multiply takes arguments x (a BigInt)
     /// and y (a BigInt) and returns a BigInt.
-    pub(crate) fn multiply(agent: &mut Agent, x: BigInt, y: BigInt) -> BigInt {
+    pub(crate) fn multiply(agent: &mut Agent, x: BigInt, y: BigInt) -> JsResult<BigInt> {
         match (x, y) {
             (BigInt::SmallBigInt(x), BigInt::SmallBigInt(y)) => {
                 let (x, y) = (x.into_i64() as i128, y.into_i64() as i128);
                 let result = x * y;
```
Comment (andreubotella) [inline nova_vm/src/ecmascript/types/language/bigint.rs:239]: **nitpick:** This should be guaranteed to not overflow, but I'd make extra sure.
```suggestion
                let result = x.checked_mul(y).unwrap();
```

File: nova_vm/src/ecmascript/types/language/bigint.rs [hunk 2]
```diff
@@ -232,35 +232,144 @@ impl BigInt {
...
+            }
+        }
+    }
+
+    /// ### [BigInt::add ( x, y )](https://tc39.es/ecma262/#sec-numeric-types-bigint-add)
+    pub(crate) fn add(agent: &mut Agent, x: BigInt, y: BigInt) -> JsResult<BigInt> {
+        match (x, y) {
+            (BigInt::SmallBigInt(x), BigInt::SmallBigInt(y)) => {
+                let (x, y) = (x.into_i64() as i128, y.into_i64() as i128);
+                let result = x + y;
```
Comment (andreubotella) [inline nova_vm/src/ecmascript/types/language/bigint.rs:269]: **needs (nonblocking):** Adding `SmallBigInt`s together should never overflow past the limit of `i64`.
```suggestion
                let (x, y) = (x.into_i64(), y.into_i64());
                let result = x.checked_add(y).unwrap();
```

File: nova_vm/src/ecmascript/types/language/bigint.rs [hunk 3]
```diff
@@ -232,35 +232,144 @@ impl BigInt {
...
+                    Ok(agent.heap.create(BigIntHeapData {
+                        data: result.into(),
+                    }))
+                }
+            }
+            (BigInt::SmallBigInt(x), BigInt::BigInt(y))
+            | (BigInt::BigInt(y), BigInt::SmallBigInt(x)) => {
+                let x = x.into_i64();
+                let y = &agent[y];
+                Ok(agent.heap.create(BigIntHeapData { data: x + &y.data }))
```
Comment (andreubotella) [inline nova_vm/src/ecmascript/types/language/bigint.rs:283]: **issue:** The result of this addition could fit in the `SmallBigInt` range. Same for the `(BigInt, BigInt)` case.

File: nova_vm/src/ecmascript/types/language/bigint.rs [hunk 4]
```diff
@@ -232,35 +232,144 @@ impl BigInt {
...
+            }
+        }
+    }
+
+    /// ### [6.1.6.2.8 BigInt::subtract ( x, y )](https://tc39.es/ecma262/#sec-numeric-types-bigint-subtract)
+    pub(crate) fn subtract(agent: &mut Agent, x: BigInt, y: BigInt) -> JsResult<BigInt> {
+        match (x, y) {
+            (BigInt::SmallBigInt(x), BigInt::SmallBigInt(y)) => {
+                let (x, y) = (x.into_i64() as i128, y.into_i64() as i128);
+                let result = x - y;
```
Comment (andreubotella) [inline nova_vm/src/ecmascript/types/language/bigint.rs:299]: **needs (nonblocking):** This subtraction shouldn't overflow `i64` either.

File: nova_vm/src/ecmascript/types/language/bigint.rs [hunk 5]
```diff
@@ -232,35 +232,144 @@ impl BigInt {
...
+                } else {
+                    Ok(agent.heap.create(BigIntHeapData {
+                        data: result.into(),
+                    }))
+                }
+            }
+            (BigInt::SmallBigInt(x), BigInt::BigInt(y)) => {
+                let x = x.into_i64();
+                let y = &agent[y];
+                Ok(agent.heap.create(BigIntHeapData { data: x - &y.data }))
```
Comment (andreubotella) [inline nova_vm/src/ecmascript/types/language/bigint.rs:312]: **issue:** This result, as well as all others in the remainder of this method, could fit in the `SmallBigInt` range.

File: nova_vm/src/ecmascript/types/language/bigint.rs [hunk 6]
```diff
@@ -232,35 +232,144 @@ impl BigInt {
...
+        match (x, y) {
+            (BigInt::SmallBigInt(x), BigInt::SmallBigInt(y)) => {
+                if y == SmallBigInt::zero() {
+                    return Err(agent.throw_exception_with_static_message(
+                        ExceptionType::RangeError,
+                        "Division by zero",
+                    ));
+                }
+                let (x, y) = (x.into_i64() as i128, y.into_i64() as i128);
+                let result = x / y;
```
Comment (andreubotella) [inline nova_vm/src/ecmascript/types/language/bigint.rs:339]: **needs (nonblocking):** A division of two integers will never be smaller in absolute value than the largest of them, so there's not only no chance of overflowing into `i128`, but the result should always be a `SmallBigInt`.

File: nova_vm/src/ecmascript/types/language/bigint.rs [hunk 7]
```diff
@@ -232,35 +232,144 @@ impl BigInt {
...
+                } else {
+                    Ok(agent.heap.create(BigIntHeapData {
+                        data: result.into(),
+                    }))
+                }
+            }
+            (BigInt::SmallBigInt(x), BigInt::BigInt(y)) => {
+                let x = x.into_i64();
+                let y = &agent[y];
+                Ok(agent.heap.create(BigIntHeapData { data: x / &y.data }))
```
Comment (andreubotella) [inline nova_vm/src/ecmascript/types/language/bigint.rs:352]: **issue:** In the cases where either `x` or `y` is a `SmallBigInt`, the result will also be a `SmallBigInt`. In the other case, the result might still be a `SmallBigInt`.

File: nova_vm/src/ecmascript/types/language/bigint.rs [hunk 8]
```diff
@@ -278,6 +387,53 @@ impl BigInt {
...
+        match (n, d) {
+            (BigInt::SmallBigInt(n), BigInt::SmallBigInt(d)) => {
+                if d == SmallBigInt::zero() {
+                    return Err(agent.throw_exception_with_static_message(
+                        ExceptionType::RangeError,
+                        "Division by zero",
+                    ));
+                }
+                let (n, d) = (n.into_i64() as i128, d.into_i64() as i128);
+                let result = n % d;
```
Comment (andreubotella) [inline nova_vm/src/ecmascript/types/language/bigint.rs:401]: **needs (nonblocking):** The result of the remainder will always have smaller absolute value than the denominator, so the there's not only no chance of overflowing into `i128`, but the result should always be a `SmallBigInt`.

File: nova_vm/src/ecmascript/types/language/bigint.rs [hunk 9]
```diff
@@ -278,6 +387,53 @@ impl BigInt {
...
+                } else {
+                    Ok(agent.heap.create(BigIntHeapData {
+                        data: result.into(),
+                    }))
+                }
+            }
+            (BigInt::SmallBigInt(n), BigInt::BigInt(d)) => {
+                let n = n.into_i64();
+                let d = &agent[d];
+                Ok(agent.heap.create(BigIntHeapData { data: n % &d.data }))
```
Comment (andreubotella) [inline nova_vm/src/ecmascript/types/language/bigint.rs:414]: **issue:** This result, as well as the `(BigInt, BigInt)` one, could fit in the `SmallBigInt` range.

File: nova_vm/src/ecmascript/types/language/bigint.rs [hunk 10]
```diff
@@ -278,6 +387,53 @@ impl BigInt {
...
+            (BigInt::BigInt(n), BigInt::SmallBigInt(d)) => {
+                if d == SmallBigInt::zero() {
+                    return Err(agent.throw_exception_with_static_message(
+                        ExceptionType::RangeError,
+                        "Division by zero",
+                    ));
+                }
+                let n = &agent[n];
+                let d = d.into_i64();
+                Ok(agent.heap.create(BigIntHeapData { data: &n.data % d }))
```
Comment (andreubotella) [inline nova_vm/src/ecmascript/types/language/bigint.rs:425]: **issue:** This result should always fit in the `SmallBigInt` range.

**Review by andreubotella — COMMENTED** [review]
**nitpick:** I wouldn't return a `JsResult` in methods that will always succeed, but if you think that's better for consistency, that's fine.

File: nova_vm/src/ecmascript/types/language/bigint.rs [hunk 11]
```diff
@@ -232,35 +232,144 @@ impl BigInt {
...
+                let (x, y) = (&agent[x], &agent[y]);
+                Ok(agent.heap.create(BigIntHeapData {
+                    data: &x.data + &y.data,
+                }))
+            }
+        }
+    }
+
+    /// ### [6.1.6.2.8 BigInt::subtract ( x, y )](https://tc39.es/ecma262/#sec-numeric-types-bigint-subtract)
+    pub(crate) fn subtract(agent: &mut Agent, x: BigInt, y: BigInt) -> JsResult<BigInt> {
```
Comment (aapoalas) [inline nova_vm/src/ecmascript/types/language/bigint.rs:295]: **issue:** Infallible functions should not return `JsResult`.

File: nova_vm/src/ecmascript/types/language/bigint.rs [hunk 12]
```diff
@@ -232,35 +232,144 @@ impl BigInt {
...
+                Ok(agent.heap.create(BigIntHeapData {
                     data: &x.data * &y.data,
-                })
+                }))
+            }
+        }
+    }
+
+    /// ### [BigInt::add ( x, y )](https://tc39.es/ecma262/#sec-numeric-types-bigint-add)
+    pub(crate) fn add(agent: &mut Agent, x: BigInt, y: BigInt) -> JsResult<BigInt> {
```
Comment (aapoalas) [inline nova_vm/src/ecmascript/types/language/bigint.rs:265]: **issue:** Infallible functions should not return a `JsResult`.

File: nova_vm/src/ecmascript/types/language/bigint.rs [hunk 13]
```diff
@@ -232,35 +232,144 @@ impl BigInt {
     ///
     /// The abstract operation BigInt::multiply takes arguments x (a BigInt)
     /// and y (a BigInt) and returns a BigInt.
-    pub(crate) fn multiply(agent: &mut Agent, x: BigInt, y: BigInt) -> BigInt {
+    pub(crate) fn multiply(agent: &mut Agent, x: BigInt, y: BigInt) -> JsResult<BigInt> {
```
Comment (aapoalas) [inline nova_vm/src/ecmascript/types/language/bigint.rs:235]: **issue:** Infallible functions should not return a `JsResult`.

**Review by aapoalas — CHANGES_REQUESTED** [review]
Good feedback from Andreu. Not much for me to add.

File: nova_vm/src/ecmascript/types/language/bigint.rs (hunk 7 above)
Comment (aapoalas) [reply nova_vm/src/ecmascript/types/language/bigint.rs]: I don't think this is true: eg. `123259032485034956830945802349830495830945890345n / 2n` will still be a heap BigInt.

**Review by aapoalas — APPROVED** [review]

**Pull request merged**